pkg/controller/service_instance.go

```diff
@@ -0,0 +1,183 @@
+package controller
+
+import (
+	"bytes"
+	"crypto/sha256"
+	"encoding/hex"
+
+	"github.com/atlassian/smith"
+	"github.com/atlassian/smith/pkg/util"
+	sc_v1b1 "github.com/kubernetes-incubator/service-catalog/pkg/apis/servicecatalog/v1beta1"
+	"github.com/pkg/errors"
+	"golang.org/x/crypto/bcrypt"
+	core_v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	unstructured_conversion "k8s.io/apimachinery/pkg/conversion/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+)
+
+const (
+	secretParametersChecksumAnnotation = smith.Domain + "/secretParametersChecksum"
+	disabled                           = "disabled"
+)
+
+// works around the fact that service catalog does not know when to send an update request if the
+// parameters section would change as the result of changing a secret referenced in parametersFrom. To
+// do this we record the contents of all referenced secrets in an annotation, compare that annotations value
+// each time a service instance is processed, and force UpdateRequests to a higher value when those secrets
+// change to trigger service catalog to send an update request.
+func (st *resourceSyncTask) forceServiceInstanceUpdates(spec *unstructured.Unstructured, actual runtime.Object, namespace string) (specRet *unstructured.Unstructured, e error) {
+	gvk := spec.GroupVersionKind()
+	gk := spec.GroupVersionKind().GroupKind()
+
+	if gk.Kind == "" || gvk.Version == "" {
+		return spec, nil
+	}
+
+	if gk.Group == sc_v1b1.GroupName && spec.GetKind() == "ServiceInstance" {
+		return st.processServiceInstance(spec, actual, namespace)
+	}
+
+	return spec, nil
+}
+
+// works by adding using an annotation to remember the previous contents of all the secrets referenced in
+// parametersFrom and forcing updateRequest to higher value when the referenced secrets change. The annotation
+// stores the hex value of the bcryted hash of the sha256 hash of the contents of all secrets referenced in
+// parametersFrom.
+//
+// The annotation is only ever added to the spec object not the actual object. The modified spec is treated as
```

Comment:
This all sounds overcomplicated - why not just always add the annotation, overwriting any existing values?